docs/customization/plugin-authors-guide.md

```diff
@@ -250,12 +250,209 @@ kaleidoscope::plugin::MyPlugin;
+    if (event.key == Key_H && keyToggledOn(event.state)) {
+	  event.key = LSHIFT(Key_H);
+	  Runtime.handleKeyEvent(event);
+
+      event.key = Key_I;
+	  Runtime.handleKeyEvent(event);
+
+	  event.key = LSHIFT(Key_1);
+    }
+    return EventHandlerResult::OK;
+  }
+```
+
+Note that, with this version, we've only sent two extra events, then changed the `event.key` value, and returned `OK` instead of `ABORT`.  This is basically the same as the above pluging that turned `Y` into `X`, but with two extra events sent first.
+
+As one extra precaution, it would be wise to mark the generated event(s) as "injected" to let other plugins know that these events should be ignored.  This is a convention that is used by many existing Kaleidoscope plugins.  We do this by setting the `INJECTED` bit in the `event.state` variable:
+
+```c++
+  EventHandlerResult onKeyEvent(KeyEvent &event) {
+    if (event.key == Key_H && keyToggledOn(event.state)) {
+      event.state |= INJECTED;
+
+      event.key = LSHIFT(Key_H);
+	  Runtime.handleKeyEvent(event);
+
+      event.key = Key_I;
+	  Runtime.handleKeyEvent(event);
+
+	  event.key = LSHIFT(Key_1);
+    }
+    return EventHandlerResult::OK;
+  }
+```
+
+If we wanted to be especially careful, we could also add the corresponding release events:
+
+```c++
+  EventHandlerResult onKeyEvent(KeyEvent &event) {
+    if (event.key == Key_H && keyToggledOn(event.state)) {
+      event.key = LSHIFT(Key_H);
+      event.state = INJECTED | IS_PRESSED;
+	  Runtime.handleKeyEvent(event);
+	  event.state = INJECTED | WAS_PRESSED;
+	  Runtime.handleKeyEvent(event);
+
+      event.key = Key_I;
+      event.state = INJECTED | IS_PRESSED;
+	  Runtime.handleKeyEvent(event);
+	  event.state = INJECTED | WAS_PRESSED;
+	  Runtime.handleKeyEvent(event);
+
+	  event.key = LSHIFT(Key_1);
+      event.state = INJECTED | IS_PRESSED;
+    }
+    return EventHandlerResult::OK;
+  }
+```
+
+### Avoiding infinite loops
+
+One very important consideration for any plugin that calls `Runtime.handleKeyEvent()` from an `onKeyEvent()` handler is recursion.  `Runtime.handleKeyEvent()` will call all plugins' `onKeyEvent()` handlers, including the one that generated the event.  Therefore, we need to take some measures to short-circuit the resulting recursive call so that our plugin doesn't cause an infinite loop.
+
+Suppose the example plugin above was changed to type the string `hi!` instead of `Hi!`.  When sending the first generated event, with `event.key` set to `Key_H`, our plugin would recognize that event as one that should be acted on, and make another call to `Runtime.handleKeyEvent()`, which would again call `MyPlugin.onKeyEvent()`, and so on until the MCU ran out of memory on the stack.
+
+The simplest mechanism used by many plugins that mark their generated events "injected" is to simply ignore all generated events:
+
+```c++
+  EventHandlerResult onKeyEvent(KeyEvent &event) {
+    if ((event.state | INJECTED) != 0)
```

Comment:
```suggestion
    if ((event.state & INJECTED) != 0)
```

surely? There should be a helper for this really I think, so I opened #1082.

Reply:
Clearly an error, thanks!